src/testing/__tests__/jobState.test.ts

```diff
@@ -0,0 +1,120 @@
+import { Entity, Relationship } from '../../framework';
+
+import { MockJobState, createMockJobState } from '../jobState';
+
+describe('entities', () => {
+  const inputEntities: Entity[] = [
+    {
+      _type: 'test_a',
+      _class: 'Resource',
+      _key: 'a',
+    },
+    {
+      _type: 'test_a',
+      _class: 'Resource',
+      _key: 'b',
+    },
+    {
+      _type: 'test_b',
+      _class: 'Resource',
+      _key: 'c',
+    },
+  ];
+
+  test('creates a job state object that can collect and query entities', async () => {
+    expect.hasAssertions();
+    const jobState = createMockJobState();
+    await jobState.addEntities(inputEntities);
+
+    await assertEntityFilteringCapabilities(jobState);
+
+    expect(jobState.collectedEntities).toEqual(inputEntities);
+  });
+
+  test('does not include pre-existing entities in collectedEntities', async () => {
+    expect.hasAssertions();
+    const jobState = createMockJobState({ entities: inputEntities });
+    await assertEntityFilteringCapabilities(jobState);
+
+    expect(jobState.collectedEntities).toEqual([]);
+  });
+
+  async function assertEntityFilteringCapabilities(jobState: MockJobState) {
+    const entitiesA: Entity[] = [];
+    await jobState.iterateEntities({ _type: 'test_a' }, (e) => {
+      entitiesA.push(e);
+    });
+
+    const entitiesB: Entity[] = [];
+    await jobState.iterateEntities({ _type: 'test_b' }, (e) => {
+      entitiesB.push(e);
+    });
+
+    expect(entitiesA).toEqual([inputEntities[0], inputEntities[1]]);
+    expect(entitiesB).toEqual([inputEntities[2]]);
+  }
+});
+
+describe('relationships', () => {
+  const inputRelationships: Relationship[] = [
+    {
+      _type: 'test_a',
+      _class: 'HAS',
+      _key: 'a|test|b',
```

Comment:
I just want to say thanks for keeping the pipe in the `_key`. I'm jarred whenever I see anything else.